ktesting/testinglogger.go

```diff
@@ -113,6 +274,32 @@ func (l *tlogger) log(what, msg string, buffer *bytes.Buffer) {
 		args = append(args, string(buffer.Bytes()[1:]))
 	}
 	l.t.Log(args...)
+
+	l.buffer.mutex.Lock()
+	defer l.buffer.mutex.Unlock()
+
+	// Store as text.
+	l.buffer.text.WriteString(string(what))
+	for i := 1; i < len(args); i++ {
+		l.buffer.text.WriteByte(' ')
+		l.buffer.text.WriteString(args[i].(string))
+	}
+	lastArg := args[len(args)-1].(string)
+	if lastArg[len(lastArg)-1] != '\n' {
+		l.buffer.text.WriteByte('\n')
+	}
+
+	// Store as raw data.
+	l.buffer.log = append(l.buffer.log,
+		LogEntry{
+			Timestamp: time.Now(),
+			Type:      what,
+			Prefix:    l.prefix,
+			Message:   msg,
+			Err:       err,
+			KVList:    append(l.values, kvList...),
```

Comment:
You probably refer to these lines from the example output, right?

```
        // log entry #0: ERROR  I failed [what something] err=failure
	// log entry #1: INFO  hello world [request 1] err=<nil>
	// log entry #2: INFO example with name [] err=<nil>
```

That's not what `go test -v` will show. What it will show are the lines above those:
```
        // ERROR I failed err="failure" what="something"
	// INFO hello world request=1
	// INFO example: with name
```

That is as readable (or unreadable) as the normal klog formatting of key/value lists.

Should I change the way how the example dumps the struct to make it less confusing?

Reply:
`entry.KVList` has multiple entries. `[what something]` is simply how '%+v` formats slices.

I think that is fine here. Tests using LogEntry are not expected to dump the content. Instead, they'll probably inspect individual fields and compare them against the expected outcome, with custom error messages in case of failures.

We could add a `LogEntry.String` pretty-printer which formats like the text output, but then how would it capture fields that are not normally included in that output, like the level? I think it's better to leave that to tests using LogEntry.

Coming back to example_test.go: I completely missed testing of the Verbosity field. I made the same mistake in the code *and* in the test. This is not unusual when developers write their own tests because they often only test what they considered during development. Reviewers need to be on the watch-out for insufficient test coverage. It's perfectly fine to ask for more tests and/or point out gaps.

To make this more obvious, I switch the dumping of LogEntry to a single `%+v`. The varying time stamp must be cleared first because `go test` expects a literal match for the full output.

Reply:
> instead, they'll probably inspect individual fields and compare them against the expected outcome,

I was precisely talking about this same behaviour.  Say you want to validate a key pair from output. Then the current one would force you to validate it a little differently and can give false positives.  Right now, the keys and values are individual items in the `KVList` instead can we combine them into a pairs as a best case effort. and store them as `k=v` in one item. 

`.WithValue("key1", "value", "key2, "value")` will not be stored as `[]{"key1", "value", "key2", "value"}`
Instead can this be turned into `[]{"key1=value", "key2=value"}` this will also help assert the outputs a bit better in the logs. 

```golang
for i := 0; i < len(kvList); i += 2 {
    end := i + 2
    if end > len(kvList) {
	end = len(kvList)
    }
     chunks = append(kvList, strings.Join(slice[i:end], "="))
}
```

That said, Let me re-iterate.

> What I am suggesting is only a nice to have if that doesn't cause too much of performance overhead to do that.

If you think this will be an overkill and a not worth adding it right now, I am find to table this discussion for a later time. 

> we could add a LogEntry.String pretty-printer which formats like the text output, but then how would it capture fields that are not normally included in that output, like the level? 

Think this would over complicate the logger than it needs to be. We have a simple `.String()` already go get the actual strings. 

> To make this more obvious, I switch the dumping of LogEntry to a single %+v. 

I noticed. This is a nice change. Asserts everything that the entry is supposed to have in one go.

Reply:
I see what you mean. But what if the log call was broken, say `logger.Info("foo", "key" /* no value here */)`? What should be in the LogEntry when its KV type requires a value?

I prefer to not answer that in the testing logger and instead let the test deal with interpreting the key/value list. Perhaps the code is already doing too much: instead of concatenating the stored key/value pairs and the parameters, having two fields in LogEntry would be better. Then the test can validate that a certain key/value pair was passed as intended.

Reply:
> What should be in the LogEntry when its KV type requires a value?

Was thinking this would be same as what the `klog` did. `key=(MISSING)`. 

> I prefer to not answer that in the testing logger and instead let the test deal with interpreting the key/value list.

Ack. I am fine with that. Thanks for the detailed discussion 🥂

Reply:
Should I concatenate (current code) or use two different fields (`WithKVList`, `ParameterKVList`)?

Reply:
@pohly  I think we can do that outside of this PR and do it in a different PR? Let us not hold this change for any longer than required. WDYT ?